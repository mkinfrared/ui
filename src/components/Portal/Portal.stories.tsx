import { Meta } from "@storybook/react";
import React, { useRef } from "react";

import Heading from "components/Heading";
import Text from "components/Text";

import { Portal } from "./Portal";
import css from "./Story.module.scss";

export default {
  title: "UI/Portal",
  component: Portal,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const children = (
  <Text>
    Lorem ipsum dolor sit amet, consectetur adipisicing elit. Repellendus
    veniam, veritatis? A adipisci, alias amet beatae cupiditate dolor doloribus
    eos esse facere, illum in inventore ipsa itaque laborum magnam nam omnis
    optio perferendis praesentium quasi recusandae repellendus repudiandae
    similique soluta sunt ullam vitae voluptate! Alias enim exercitationem
    molestiae? Cupiditate earum expedita illum, ipsum iste minus odit provident
    voluptas! Assumenda commodi consequuntur culpa cumque doloremque eum
    laudantium nisi quidem quis rem sapiente sunt, suscipit temporibus, totam ut
    velit voluptate voluptatem. Assumenda doloribus expedita facilis impedit
    laborum natus quidem temporibus voluptatibus? Aliquid dolor earum eius
    excepturi expedita omnis porro provident quibusdam soluta!
  </Text>
);

const ReplaceHtml = () => {
  const ref = useRef<HTMLDivElement | null>(null);

  return (
    <div className={css.Story}>
      <div ref={ref}>
        <Heading variant="h3">
          This block will be replaced by Portal's children
        </Heading>
      </div>
      <div>
        <Heading variant="h3">Simple block</Heading>
        <Portal container={ref}>{children}</Portal>
      </div>
    </div>
  );
};

const NoReplaceHtml = () => {
  const ref = useRef<HTMLDivElement | null>(null);

  return (
    <div className={css.Story}>
      <div ref={ref}>
        <Heading variant="h3">This block will stay after Portal render</Heading>
      </div>
      <div>
        <Heading variant="h3">Simple block</Heading>
        <Portal replaceHtml={false} container={ref}>
          {children}
        </Portal>
      </div>
    </div>
  );
};

ReplaceHtml.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { ReplaceHtml, NoReplaceHtml };
