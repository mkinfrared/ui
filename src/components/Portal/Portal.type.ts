import React from "react";

export type PortalProps = {
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  children?: React.ReactNode;
  container: React.RefObject<HTMLElement>;
  replaceHtml?: boolean;
};
