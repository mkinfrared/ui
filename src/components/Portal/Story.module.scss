.Story {
  div {
    padding: 40px;

    &:first-of-type {
      background-color: paleturquoise;
    }

    &:last-of-type {
      background-color: palegoldenrod;
    }
  }
}
