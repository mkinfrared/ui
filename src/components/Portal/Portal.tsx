import { memo, useEffect, useState } from "react";
import { createPortal } from "react-dom";

import { PortalProps } from "./Portal.type";

const Portal = ({ children, container, replaceHtml = true }: PortalProps) => {
  const [htmlElement, setHtmlElement] = useState(container.current);

  useEffect(() => {
    const { current } = container;

    if (current && replaceHtml && !htmlElement) {
      current.innerHTML = "";
    }

    setHtmlElement(current);
  }, [container, htmlElement, replaceHtml]);

  if (!htmlElement) {
    return null;
  }

  if (replaceHtml && !htmlElement) {
    return null;
  }

  return createPortal(children, htmlElement);
};

export { Portal };

export default memo(Portal);
