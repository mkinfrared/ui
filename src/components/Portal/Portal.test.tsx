import { render } from "@testing-library/react";
import React, { createRef } from "react";

import { Portal } from "./Portal";

describe("<Portal />", () => {
  it("should be defined", () => {
    expect(Portal).toBeDefined();
  });

  it("should match the snapshot", () => {
    const ref = createRef<HTMLDivElement>();

    const { container } = render(
      <div ref={ref}>
        <p>lorem ipsum</p>
        <Portal container={ref}>
          <p>marklar</p>
        </Portal>
      </div>,
    );

    expect(container).toMatchSnapshot();
  });
});
