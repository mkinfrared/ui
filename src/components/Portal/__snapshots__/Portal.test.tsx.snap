// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portal /> should match the snapshot 1`] = `
<div>
  <div>
    <p>
      marklar
    </p>
  </div>
</div>
`;
